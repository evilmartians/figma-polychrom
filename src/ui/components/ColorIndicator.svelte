<script lang="ts">
  import ColorPreview from '~ui/components/ColorPreview.svelte';
  import { copy } from '~ui/services/use-copy-to-clipboard/useCopyToClipboard.ts';
  import {
    getFormatterForCSS,
    getFormatterForDisplaying,
  } from '~utils/colors/formatters.ts';
  import { type Oklch } from 'culori/fn';

  import { colorSpaceDisplayMode } from '../stores/color-space-display-mode.ts';
  import Tooltip from './Tooltip.svelte';

  interface ColorIndicatorProps {
    borderColor: string;
    fill: { hex: string; oklch: Oklch };
    indicatorColor: string;
    isBlended: boolean;
    textColor: string;
  }

  const {
    borderColor,
    fill,
    indicatorColor,
    isBlended,
    textColor,
  }: ColorIndicatorProps = $props();

  const formatColorForDisplay = $derived(
    getFormatterForDisplaying($colorSpaceDisplayMode)
  );
  const displayValue = $derived(formatColorForDisplay(fill.oklch));

  const formatColorForCSS = getFormatterForCSS($colorSpaceDisplayMode);
  const cssValue = formatColorForCSS(fill.oklch);
  let copied = $state('');
  const onCopy = (): void => {
    copied = cssValue;
    setTimeout(() => {
      copied = '';
    }, 2000);
  }
</script>

<Tooltip>
  {#snippet trigger()}
      <button
        style:color="{indicatorColor}"
        class="interactive"
        type="button"
        use:copy
        oncopied={onCopy}
      >
        <div
          class="flex items-center rounded-[7px] p-1 hover:bg-indicatorsHover active:bg-indicatorsActive"
        >
          <div class="mr-2">
            <ColorPreview {borderColor} {indicatorColor} {isBlended} />
          </div>
          <span style:color="var({textColor})">{displayValue}</span>
        </div>
      </button>
  {/snippet}

  {#if copied}
    Copied!
  {:else}
    <span class="whitespace-nowrap">Copy as CSS</span>
  {/if}
</Tooltip>
