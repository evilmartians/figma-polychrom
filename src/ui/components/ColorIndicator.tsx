import { useStore } from '@nanostores/preact';
import { ColorPreview } from '~ui/components/ColorPreview.tsx';
import { ThemeVariablesKeys } from '~ui/components/ThemeVariablesProvider.tsx';
import { Tooltip } from '~ui/components/Tooltip.tsx';
import {
  getFormatterForCSS,
  getFormatterForDisplaying,
} from '~utils/colors/formatters.ts';
import { useClipboard } from '~utils/use-clickboard.ts';
import { type Oklch } from 'culori/fn';
import { type JSX } from 'preact';

import { $colorSpaceDisplayMode } from '../stores/color-space-display-mode.ts';

interface ColorIndicatorProps {
  borderColor: string;
  fill: { hex: string; oklch: Oklch };
  indicatorColor: string;
  isBlended: boolean;
  textColor: string;
}

export const ColorIndicator = ({
  borderColor,
  fill,
  indicatorColor,
  isBlended,
  textColor,
}: ColorIndicatorProps): JSX.Element => {
  const colorSpaceDisplayMode = useStore($colorSpaceDisplayMode);

  const formatColorForDisplay = getFormatterForDisplaying(
    colorSpaceDisplayMode
  );
  const displayValue = formatColorForDisplay(fill.oklch);

  const formatColorForCSS = getFormatterForCSS(colorSpaceDisplayMode);
  const cssValue = formatColorForCSS(fill.oklch);

  const [isCopied, setCopied] = useClipboard(cssValue, {
    successDuration: 2000,
  });

  return (
    <Tooltip content={isCopied ? 'Copied!' : 'Copy as CSS'}>
      <button
        style={{
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          '--indicators-active': `var(${ThemeVariablesKeys.secondary24})`,
          '--indicators-hover': `var(${ThemeVariablesKeys.secondary16})`,
          color: indicatorColor,
        }}
        className="interactive"
        onClick={setCopied}
        type="button"
      >
        <div className="flex items-center rounded-[7px] p-1 hover:bg-indicatorsHover active:bg-indicatorsActive">
          <div className="mr-2">
            <ColorPreview
              borderColor={borderColor}
              indicatorColor={indicatorColor}
              isBlended={isBlended}
            />
          </div>
          <span style={{ color: `var(${textColor})` }}>{displayValue}</span>
        </div>
      </button>
    </Tooltip>
  );
};
