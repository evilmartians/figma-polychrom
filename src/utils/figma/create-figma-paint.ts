import { converter, formatHex } from 'culori';

import { type FigmaPaint } from '../../types/common.ts';

const convertToOklch = converter('oklch');

export const createFigmaPaint = (
  color: RGB,
  opacity: number = 1
): FigmaPaint => {
  return {
    color,
    hex: formatHex({ ...color, mode: 'rgb' }),
    oklch: convertToOklch({ ...color, mode: 'rgb' }),
    opacity,
    type: 'SOLID',
  };
};
