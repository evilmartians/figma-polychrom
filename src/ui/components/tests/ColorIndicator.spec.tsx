import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { createFigmaPaint } from '../../../utils/figma/create-figma-paint.ts';
import { ColorIndicator } from '../ColorIndicator';

describe('ColorIndicator component', () => {
  test('Render without crash', () => {
    expect(
      render(
        <ColorIndicator
          fill={createFigmaPaint({ b: 0.5, g: 0.5, r: 0.5 }, 1)}
        />
      ).container
    ).toMatchSnapshot();
  });
});
