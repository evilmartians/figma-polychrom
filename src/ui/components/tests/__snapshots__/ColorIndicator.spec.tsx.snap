// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorIndicator component > Render without crash 1`] = `
<div>
  <svg
    fill="none"
    height="26"
    viewBox="0 0 22 26"
    width="22"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <path
        d="M0 6.5L11 0L22 6.5V19.5L11 26L0 19.5V6.5Z"
        fill="#808080"
        fill-opacity="1"
      />
      <path
        d="M0.25 19.3573V6.64266L11 0.290385L21.75 6.64266V19.3573L11 25.7096L0.25 19.3573Z"
      />
    </g>
  </svg>
</div>
`;
